### DISCUSSION TITLE
ESP32-C3 - The connected target does not have support for nanoFramework.Hardware.Esp32.Rmt

### DISCUSSION THREAD

**Discussion by tank104** [nanoframework/Home#1198] [Q&A]
Hi,
I am using ESP32-C3-DevKitC-02 Development Board, and trying to change the LED light on the board.
From reading it looks this is and addressable” RGB LED.
Googling has shown me that I need to use nanoFramework.Hardware.Esp32.Rmt.

I have shown code below.

However when I deploy this I get error:
[The connected target does not have support for nanoFramework.Hardware.Esp32.Rmt.]

Do I need to flash the board with extra libraries to support these packages?  I can't find anything else to flash that indeicates how to make this work?

Any help appreciated - have searched as much as I can for a solution but come up with nothing.

`      s_GpioController = new GpioController();
      GpioPin led = s_GpioController.OpenPin(8, PinMode.Output);
      using (var rmt_transmitter = new TransmitterChannel(8))
      {
        // transmitter configuration
        rmt_transmitter.CarrierEnabled = false;
        //rmt_transmitter.isSource80MHz = true;
        rmt_transmitter.ClockDivider = 4; // base period 80 MHz / 4 => 20 MHz -> 0.05 us
        //rmt_transmitter.IsChannelIdle = true;
        //rmt_transmitter.TransmitIdleLevel = false;

        rmt_transmitter.AddCommand(new RmtCommand(100, true, 100, false));
        rmt_transmitter.AddCommand(new RmtCommand(100, true, 100, false));
        rmt_transmitter.AddCommand(new RmtCommand(100, true, 100, false));
        rmt_transmitter.AddCommand(new RmtCommand(100, true, 100, false));
        // create new pulse command for to transmitt
        // IDLE  0.5us  1us       1.5 us           2 us             2.5 us	   IDLE..
        //      |----|        |------------|                |--------------------|
        // -----|    |--------|            |----------------|                    |------
        rmt_transmitter.Send(true);
      }`

**Comment by tank104** [comment]
I am also getting a similiar issue when trying to use I2s:

The connected target does not have support for System.Device.I2s

**Reply by josesimoes** [reply]
This is happening because support for ESP32C3 is a bit recent and those 2 features aren't enabled yet as they require some coding.

**Reply by tank104** [reply]
Ahh ok thanks - does it typically take long to get added.  Or is it something I can do?  (very good c# coder, but near zero experience in nano - if so is there a guide?)

I guess I have brought the latest chips :(

**Reply by josesimoes** [reply]
Really depends on someone tackling those tasks... None of it it's particularly challenging, yet, this is run mostly out of community effort.
As for the coding: I'm afraid these are at the native end of it, meaning C/C++.

Nevertheless, let's not waste your enthusiasm and willing to help! 😄 Can I interest you on one of the open issues? There are numerous ones C# related. Some are labeled as "good first issue" 😉   This could be a smooth way for getting into this.

**Reply by MrCSharp22** [reply]
Hi @josesimoes 
Is there a write up somewhere on how to bring the RMT library to the ESP32 C3 target? I'd like to tackle this but I am not familiar with CMake. So if there is such a doc, it'd be super helpful.

**Reply by josesimoes** [reply]
@MrCSharp22 there is no such doc. You're better have the compiler gide you through this.
IDF API is usually the same or pretty close between series, so I'm guessing you'll have to just perform some tweaks (if any are required).

**Reply by tank104** [reply]
@josesimoes just following up on this - is there anywhere I can see what boards have better support?

Is it safe to assume any of these devices have good support for all libraries:
https://docs.nanoframework.net/content/getting-started-guides/where-to-buy-devices.html

Might be easier for me to just buy them, instead of waiting for changes

**Comment by Ellerbach** [comment]
> Is it safe to assume any of these devices have good support for all libraries:

Yes, it is safe but it's just a subset and some suggestions.

You have the more exhaustive list here [and here](https://github.com/nanoframework/home#firmware-for-reference-boards)

**Reply by tank104** [reply]
Thanks but that list doesn’t show which devices have compatibility/support with which nanoframework libraries?  I.e i2s

**Reply by Ellerbach** [reply]
When you click expand, you have the main libraries. Now, you're right, the list is not fully up to date.
In theory RMT is part of the [build](https://github.com/nanoframework/nf-interpreter/blob/ebc9be35eb77192ab0a31401159def465e19dcf3/targets/ESP32/CMakePresets.json#L219). And I2S should be on by default as well.

Now, your code is not correct. You must not open the GPIO pin.

**Comment by josesimoes** [comment]
I2C is set to off for the C3 series, that's why the target is reporting what you're seeing.
I'm adjusting this in nanoframework/nf-interpreter#2691 so you can test if support for I2C is working properly for C3.

**Reply by tank104** [reply]
Great will do, let me know when new update ready.

can DAC be enabled too?

**Reply by tank104** [reply]
And RMT.  Thanks

**Reply by josesimoes** [reply]
The build is failing. Some work is required to enable I2C on the C3. That's the reason for it being disabled.
You can check the build failure [here](https://dev.azure.com/nanoframework/nf-interpreter/_build/results?buildId=42877&view=logs&jobId=2efa19b6-03fa-536b-b1d3-d77e11074f2a&j=2efa19b6-03fa-536b-b1d3-d77e11074f2a&t=7d528cd3-a33d-524f-8ec7-9412dceeb7c0) 
In case you're interested in fixing it, please go ahead and work on that PR.

**Reply by Ellerbach** [reply]
I had a quick look. In short: the volatile configuration should be replace by the proper API. This has changed in 4.0 and should be much easier and straight forward. If anyone wants to give a try, go ahead! It's about removing the 4 lines as the bug is solved ;-) https://github.com/nanoframework/nf-interpreter/blob/ebc9be35eb77192ab0a31401159def465e19dcf3/targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp#L325

**Reply by Ellerbach** [reply]
(note: it was a bug from Espressif)

**Reply by tank104** [reply]
If I have understood correctly - then I have made that change (removing all the REG_SET_BIT - and confirm build is working:
https://github.com/nanoframework/nf-interpreter/pull/2692

Are there instructions I can use to take this build output and test it locally with one of the chips?  Sorry haven't done it before, and do need some pointers.

**Reply by josesimoes** [reply]
@tank104 you can download the build artifacts from the pipeline [here](https://dev.azure.com/nanoframework/nf-interpreter/_build/results?buildId=42886&view=results) then use nanoff to flash the ESP32 for testing.

**Reply by tank104** [reply]
I had a look there but don't seem to be able to download the artifacts, is it a security issue?  Pretty sure I can do it fine on my own DevOp repos inthe past

https://dev.azure.com/nanoframework/nf-interpreter/_build/results?buildId=42886&view=artifacts&pathAsName=false&type=publishedArtifacts

**Reply by josesimoes** [reply]
😯  indeed!! One needs to be authenticated to download artifacts from a PR build. It seems... as I'm alwyas authenticated, I had never spotted this... pretty sure there is a permission for this... 
Do you happen to know what it is?

**Reply by tank104** [reply]
I don’t, but does this help:
https://developercommunity.visualstudio.com/t/permissions-needed-to-download-build-artifacts/578781

**Comment by tank104** [comment]
I don’t sorry, always been admin on repos so never thought about it

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
